collections/keys.go

```diff
@@ -0,0 +1,37 @@
+package collections
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"strconv"
+)
+
+// Uint64Key can be used to encode uint64 keys.
+// Encoding is big endian to retain ordering.
+var Uint64Key KeyEncoder[uint64] = uint64Key{}
+
+var errDecodeKeySize = errors.New("collections: decode error, wrong byte key size")
+
+type uint64Key struct{}
+
+func (u uint64Key) Encode(key uint64) ([]byte, error) {
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, key)
+	return b, nil
+}
+
+func (u uint64Key) Decode(b []byte) (int, uint64, error) {
+	if size := len(b); size != 8 {
+		return 0, 0, fmt.Errorf("%w: wanted 8, got: %d", errDecodeKeySize, size)
+	}
+	return 8, binary.BigEndian.Uint64(b), nil
+}
+
+func (u uint64Key) Stringify(key uint64) string {
+	return strconv.FormatUint(key, 10)
+}
+
+func (u uint64Key) KeyType() string {
+	return "uint64"
+}
```

Comment:
This was the v0 approach I pioneered and oh boy.... the pain in having to write each time MyMap.Set(ctx, collections.Uint64(someUintValue), value)... couple it with the `collections.Pair` for 2-part keys and you were doing more type conversions than actual business logic.

Reply:
You shouldn't need to do that, though.

Reply:
I think I am maybe misunderstanding how you are proposing to implement it, or rather how you're proposing to use it in let's say `Map`.

Reply:
What is the point of the Map type?

Reply:
Map is the core type of collections (which everything else builds on top of) that maps arbitrary types to arbitrary values in the KV.

Map uses KeyEncoder and ValueEncoder to encode and decode the generic type parameters K and V to and from bytes.

This allows us to define Map's type parameters K and V  as `any`. Otherwise we would need to type constraint them into types which are able to marshal/unmarshal themselves.

The advantage of keeping this separate distinction between the map API and the encoding API (defined in value and key encoders) allows us to have maps with primitive  (golang) types and allow for different encoding semantics (null terminated, length prefixed) for the same type (eg: string, sdk.AccAddress), which allows us to also retain serialisation backwards compatibility (important in order to upgrade without state migrations).

So given that the properties we need are:
- no type constrain on K and V in Map
- multiple ways to encode the same type K

Back to my question above, how would what you're proposing above work?

Reply:
If K and V are supposed to be things that can be serialized to bytes, then they are not `any` by definition. `any` does not provide that capability. Keys must be at least `comparable` and values can be anything. But yes, if you're saying that K and V are things that have well-defined serialized representations then yes, they do in fact need to have type constraints that estalibish that expectation. 

> no type constrain on K and V in Map

Factually not possible (for K at least).

> multiple ways to encode the same K

OK, but each encoding needs to be distinguishable from others at the serialization layer, which this PR does not satisfy.

Reply:
it's a path I've already explored, you constrain K to be of a well defined Key interface , lets say I create `type StringKey string`, then in all your APIs you have to explictly cast/convert your primitive golang string to `StringKey`. 
 
 I can show you the diff of the change from constrained K to unconstrained K: 
 - https://github.com/NibiruChain/nibiru/pull/991/files#diff-1cc10500f39630f207248d6d81aafb29eb3d1ba6b7b004b01af1839bbe25b539R65
 - https://github.com/NibiruChain/nibiru/pull/991/files#diff-2034071ca621e409e75a65d30fb979be15c3f76354f3e8610e9727214c1be635R191
 
 It looked good from collections POV, it sucked when you were using it. As I said the code was more type conversions than actual business logic. It was hard to read, and for more complex key types (joined ones) like `Pair` it was a total mess. 
 
Now, it's about trade-offs I guess, and the trade off here was: delegate encoding duty to a separate type (key encoder), which should be used to distinguish keys at serialisation layer (the map K type itself shouldn't, as it can have more bytes representations) and get something which is humanly usable.
 
Or take the more idiomatic path, the one you mentioned above and have your code cluttered by type conversions. 

Then also migration of the idiomatic path is exponentially more difficult.

Regarding `comparable` this is an optional constraint, not always we need comparisons between keys, and comparability in the end matters only at the bytes layer (for ordering only).

Reply:
Just to wrap my thoughts up, the constraint on K is transitively satisfied by KeyEncoder[K] which is passed as a a parameter during map's instantiation. This allows us to satisfy the following properties:

- no type constrain on K and V in Map (can use string, uint64 aka types where you cannot implement anything on top of)
- multiple ways to encode the same type K (performance, backwards compat)

A middle-ground would be to add 2 more type parameter to Map, which is the key and value encoder itself.

`Map[K any, V any, KE KeyEncoder[K], VE ValueEncoder[V]`

Reply:
this tightly couples the collection type with its expected encoding semantics

Reply:
I think the use of generics is fine the way it is. There is no need for additional constraints. Delegating the capability to the KeyEncoder and ValueEncoder types handles this.

Reply:
The good thing about having it as `Map[K any, V any, KE KeyEncoder[K], VE ValueEncoder[V]]`, is that it makes:

`Map[string, uint64, StringKey, Uint64Value]` fundamentally different from `Map[string, uint64, NullTerminatedString, Uint64Value]`.

For stateless encoders (which is 99% of cases), you wouldn't need to call `NewMap(storeKey, prefix, StringKey, Uint64Value)`. 

But just `NewMap(storeKey, prefix)`. 
For stateful encoders we could add (using the options pattern) a `KeyEncoderInit` function.
And, in the future, AccAddressKeyEncoder would be initted as `NewMap(storeKey, prefix, WithAccAddressKeyEncoder(bech32prefix))`

Reply:
@testinginprod 

> Regarding comparable this is an optional constraint, not always we need comparisons between keys, and comparability in the end matters only at the bytes layer (for ordering only).

Are we talking about Go `map` types, or something else?

Go `map` types absolutely require their keys to be `comparable`, this is not optional.

However, if we're talking about some other, abstract, concept of a map, OK! We can define a type that models that abstract-map, and that type can definitely use keys which may not satisfy Go `comparable`. But! An abstract-map cannot be represented directly as a Go-map, right? For that to work, the abstract-map would need to transform its keys from its not-necessarily-comparable type, into some definitely-comparable type that can be used by a Go-map. And, as you note, there are many possible ways to do that transformation. So there is no single Go-map representation of this abstract-map. There are many, depending on the key-transformation function. (And maybe also a value-transformation function!)

What does this mean? It means a KV collection is an abstract-map, i.e. with keys that are not Go `comparable` without transformation. And _that_ means a KV collection factually does not have a single well-defined relationship to any SDK Go-map. KV collections to Go collections are N-to-M. It's not possible to translate between them without explicily specifying a translation (mapping) function.

Those facts complicate stuff like you describe here:

> The advantage of keeping this separate distinction between the map API and the encoding API (defined in value and key encoders) allows us to have maps with primitive (golang) types and allow for different encoding semantics (null terminated, length prefixed) for the same type (eg: string, sdk.AccAddress), which allows us to also retain serialisation backwards compatibility (important in order to upgrade without state migrations).

If a remote KV collection of Map[X]Y uses some encoding E1 when translating a Go type X to a KV type X, and/or a Go type Y to a KV type Y, then that encoding E1 has to be expressed, explicitly, somewhere, as part of protocol negotiation. How does that happen here?

Reply:
> Are we talking about Go map types, or something else?

When I refer to map, I mean the collections Map. I am sorry if the name might be confusing but it's what 'others' have been using in blockchain for the past years in smart contract development languages:
- https://docs.soliditylang.org/en/v0.4.21/types.html#mappings
- https://docs.rs/cw-storage-plus/latest/cw_storage_plus/struct.Map.html
[I can make more examples if needed]

>  It's not possible to translate between them without explicitly specifying a translation (mapping) function

Yes it's absolutely correct.

> And that means a KV collection factually does not have a single well-defined relationship to any SDK Go-map

Yet again correct, and I want to mention this [`single well-defined relationship to any SDK Go-map`] is not a property we need or want.

> f a remote KV collection of Map[X]Y uses some encoding E1 when translating a Go type X to a KV type X, and/or a Go type Y to a KV type Y, then that encoding E1 has to be expressed, explicitly, somewhere, as part of protocol negotiation

If you refer to: `how do I specify my encoding` then it happens at instantiation level: https://github.com/cosmos/cosmos-sdk/pull/14134/files#diff-4bc8b2615c7a6a4fd8cd73f338d8df33eeadcb926fab317b7c1a19798e11b5efR11

If you refer to: `how do I exchange information regarding encoding to an external consumer` then it's a separate matter. 

Let's not forget that this PR/ADR aims to improve the current status quo in which every module developer has to work with `codec` and create its own `Get/Set/Iterate/Remove` methods manually for every storage object.

Ref: https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/delegation.go#L15

Reply:
> When I refer to map, I mean the collections Map

Is there a canonical definition of a collections Map?

From [cw-storage-plus/struct.Map](https://docs.rs/cw-storage-plus/latest/cw_storage_plus/struct.Map.html) I can see a set of methods, which unfortunately have no documentation or any defined semantics at all really. For example there is nothing that asserts iteration should be determinstic. But in any case it express a concept which — as you say — is basically unrelated to the Go map type. 

Cool! So we have to define a `type Map struct` or maybe `type Map interface` which satisfies this higher-level Map concept. But the Map defined here is actually super weird! It captures de/serialization functions, and a key namespace, but not actually any notion of the key or value data itself. A value of type Map isn't maintaining any state. The stateful thing is given as a parameter to Map methods like save, remove, load, has, etc. This isn't a Map at all! It's actually an adapter, a transform function, from a set of native keys and vals provided by callers, into a set of encoded keys and vals maintained in a storage layer. 

This isn't a Map, it's a Mapping!

Reply:
> That makes the type an adapter, a transform function, from a set of native keys and vals provided by callers, into a set of encoded keys and vals maintained in a storage layer. It's not a Map at all! It's like.. a Mapping! That's a fundamentally different thing.

Okay fine, so what's the action point here, rename the type? :S

Reply:
Def.

Reply:
Def stands for?

Reply:
Sorry! Definitely.

Reply:
Id prefer if we can keep a simple api that is similar to something that many people are starting to use. This is nit anyways, no need to dive into semantics to this point. resolving this comment.